docs/posts/querying-pypi-metadata-compiled-languages/index.qmd

```diff
@@ -0,0 +1,267 @@
+filetype is used in a Python project.
+
+The dataset we're using has _every file in every project_ -- what criteria should we use?
+
+We can follow Seth's lead and look for things:
+
+1. A file extension that is one of: `asm`, `cc`, `cpp`, `cxx`, `h`, `hpp`, `rs`, `go`, and variants of `F90`, `f90`, etc...
+   That is, C, C++, Assembly, Rust, Go, and Fortran.
+2. We exclude matches where the file path is within the `site-packages/` directory.
+3. We should exclude matches that are in directories used for testing.
+
+```{python}
+expr = pypi.filter(
+    [
+        _.path.re_search(r"\.(asm|c|cc|cpp|cxx|h|hpp|rs|[Ff][0-9]{0-2}(?:or)?|go)$"),
+        ~_.path.re_search(r"(^|/)test(|s|ing)"),
+        ~_.path.contains("/site-packages/"),
+    ]
+)
+```
+
+```{python}
+expr
+```
+
+That _could_ be right -- we can peak at the filename at the end of the `path` column to do a quick check:
+
+```{python}
+expr.path.split("/")[-1]
+```
+
+Ok! Next up, we want to group the matches by:
+
+1. The month that the package / file was published
+   For this, we can use the `truncate` method and ask for month as our truncation window.
+2. The file extension of the file used
+
+```{python}
+expr.group_by(
+    [
+        _.uploaded_on.truncate("M").name("month"),
+        _.path.re_extract(r"\.([a-z0-9]+)$", 1).name("ext"),
+    ]
+).aggregate()
+```
+
+That looks promising. Now we need to grab the package names that correspond to a
+given file extension in a given month and deduplicate it. And to match Seth's
+results, we'll also sort by the month in descending order:
+
+```{python}
+expr = (
+    expr.group_by(
+        [
+            _.uploaded_on.truncate("M").name("month"),
+            _.path.re_extract(r"\.([a-z0-9]+)$", 1).name("ext"),
+        ]
+    )
+    .aggregate(projects=_.project_name.collect().unique())
+    .order_by(_.month.desc())
+)
+
+expr
+```
+
+## Massage and plot
+
+Let's continue and see what our results look like.
+
+We'll do a few things:
+
+1. Combine all of the C and C++ extensions into a single group by renaming them all.
+2. Count the number of distinct entries in each group
+3. Plot the results!
+
+```{python}
+collapse_names = expr.mutate(
+    ext=_.ext.re_replace(r"cxx|cpp|cc|c|hpp|h", "C/C++")
+    .replace("rs", "Rust")
+    .replace("go", "Go")
+    .replace("asm", "Assembly"),
+)
+
+collapse_names
+```
+
+Note that now we need to de-duplicate again, since we might've had separate
+"unique" entries for both an `h` and `c` file extension, and we don't want to
+double-count!
+
+We could rewrite our original query and include the renames in the original
+`group_by` (this would be the smart thing to do), but let's push on and see if
+we can make this work.
+
+The `projects` column is now a column of string arrays, so we want to collect
+all of the arrays in each group, this will give us a "list of lists", then we'll
+want to `flatten` that list, then `unique().length()` as before.
```

Comment:
```suggestion
all of the arrays in each group, this will give us a "list of lists", then we'll
`flatten` that list and call `unique().length()` as before.
```